### DISCUSSION TITLE
auto control ego car in MultiAgentAction and MultiAgentObservation

### DISCUSSION THREAD

**Discussion by LianhaoYin** [eleurent/highway-env#249] [Ideas]
Could you help to give some instruction or help to create an ego car that can shift lanes and pass cars in the traffic automatically in multiagent action and multiagent observation settings? Especially different action type in multiagetnaction configuration @eleurent

**Comment by eleurent** [comment]
Regarding the choice of action type, I guess it kind of depends on your own goals and requirements. But generally, there is a range of choices for the action space that go from largest and most general to the smallest and most constrained.

- `ContinuousAction` is the most general type you can use, since it gives the agent full control over the vehicle (gas and steering). However, it is also the largest space, since it is continuous of dimension 2, which means you won't be able to use algorithms that only support discrete actions, and suffer a higher computational cost for some techniques such as sampling-based search. It also does not prevent the agent from going to irrelevant states, such as e.g. driving off-road or at an irrelevant speed etc., which makes exploration harder.
- `DiscreteAction` is a simple uniform discretization of the continuous space, so while it reduces its size, it can be still very large if you use a small discretization step. I guess the smallest size that makes sense would be to use 3x3=9 actions, but then you lose in accuracy. None of the exploration-related issues are solved either.
- `DiscreteMetaAction` is the smallest set of actions, with size only 5. It constrains the possible trajectories that the agent follows quite a lot, which means that you may be stuck with only suboptimal trajectories (the agent cannot change lane while changing speed for instance). On the other hand, having a smaller decision space makes the optimization problem easier to solve, and since the meta actions should all correspond to somewhat sensible behaviors (e.g. the agent never drives off-road), the agent will naturally explore a more useful part of its state space, so exploration is easier.
